lti_tool/views.py

```diff
@@ -0,0 +1,320 @@
+from exercise.models import LearningObject, BaseExercise
+from exercise.views import ExerciseView, SubmissionView
+from userprofile.models import UserProfile
+from lib.viewbase import BaseTemplateView, BaseRedirectView, BaseMixin
+from authorization.permissions import ACCESS
+from .utils import get_tool_conf, get_launch_data_storage, get_launch_url
+
+# CSRF exempts due to difficulty in proper implementation. To re-review later.
+@csrf_exempt
+@xframe_options_exempt
+def lti_login(request):
+    tool_conf = get_tool_conf()
+    launch_data_storage = get_launch_data_storage()
+
+    oidc_login = DjangoOIDCLogin(request, tool_conf, launch_data_storage=launch_data_storage)
+    target_link_uri = get_launch_url(request)
+    return oidc_login.enable_check_cookies().redirect(target_link_uri)
+
+
+class LtiLaunchView(BaseRedirectView):
+
+    access_mode = ACCESS.ANONYMOUS
+
+    @csrf_exempt
+    @xframe_options_exempt
+    def post(self, request, *args, **kwargs):
+        tool_conf = get_tool_conf()
+        launch_data_storage = get_launch_data_storage()
+        message_launch = DjangoMessageLaunch(self.request, tool_conf, launch_data_storage=launch_data_storage)
+        self.message_launch_data = message_launch.get_launch_data()
+        
+        # Get or create user
+        username = self.message_launch_data["email"]
+        try:
+            # Moodle sends username in this parameter
+            username = self.message_launch_data['https://purl.imsglobal.org/spec/lti/claim/ext']['user_username']
+        except (KeyError, TypeError):
+            # Email always received, use as fallback
+            username = self.message_launch_data['email']
+        users = User.objects.filter(username=username)
+        if len(users) == 1:
+            self.user = users[0]
+        elif len(users) == 0:
+            self.user = User.objects.create(
+                username=username,
+                email=self.message_launch_data["email"],
+                first_name=self.message_launch_data["given_name"],
+                last_name=self.message_launch_data["family_name"]
+            )
+            profile = UserProfile.objects.get(user=self.user)
+            profile.language = self.message_launch_data["https://purl.imsglobal.org/spec/lti/claim/launch_presentation"]["locale"]
+            profile.student_id = self.message_launch_data["https://purl.imsglobal.org/spec/lti/claim/lis"]["person_sourcedid"]
+            profile.save()
+
+            self.profile = profile
+        login(self.request, self.user)
+        self.request.session["lti-launch-id"] = message_launch.get_launch_id()
+
+        if message_launch.is_resource_launch():
+            launch_params = self.message_launch_data["https://purl.imsglobal.org/spec/lti/claim/custom"]
+            instance = CourseInstance.objects.get(
+                course__url=launch_params["course_slug"],
+                url=launch_params["instance_slug"]
+            )
```

Comment:
`get_object_or_404()` would be nice here.

https://docs.djangoproject.com/en/3.2/topics/http/shortcuts/#get-object-or-404